### DISCUSSION TITLE
Suggestion: always send release events for modifier keys

### DISCUSSION THREAD

**Discussion by tokyo4j** [labwc/labwc#2448] [General]
Currently, labwc usually don't send key-release events for modifier keys if they have triggered a keybind. This prevents, for example, firefox from showing its menu bar when a keybind with Alt is triggered. But since this caused issues with blender where the modifier was stuck in the client, #2377 added `wantAbsorbedModifierReleaseEvents` window rule to optionally send key-release events for modifier keys.

However, I wonder if this is a good approach. Here are the reasons:
- When a user hit the issue like blender's, he will just assume it's a bug and it's unlikely for him to find `wantAbsorbedModifierReleaseEvents` from our documentation soon.
- Other compositors just always send key-release events for modifier keys.

Therefore, I suggest always sending key-release events for modifier keys and propose alternate approaches for preventing firefox to show menu bar with Alt-* keybind:

1. Send `wl_keyboard.{leave,enter}` events to the client when a keybind with modifiers is triggered. This lets the client reset its state of pressed modifier keys so that firefox never show its menu bar when a Alt-* keybind is triggered. A downside of this approach is that the compositor sends many unnecessary `wl_keyboard.{leave,enter}` events, especially the keybind is triggered with key repetitions. For example, when the user has a keybind like `<keybind key="W-bracketRight" name.action="ZoomIn" />` and keeps pressing `Win`+`[` to zoom the screen by a large amount, the compositor continuously sends repeated `wl_keyboard.{leave,enter}` events. (EDIT: maybe I was biased to the implementation below and we can just not send `wl_keyboard.{leave,enter}` when repeated keybind executions)

   This can be tested here: https://github.com/tokyo4j/labwc/tree/refresh-keyboard-on-action

3. Clear the keyboard focus when activating window-switcher or opening a menu so that firefox doesn't show its menu bar at least with the default Alt-Tab or Alt-Space keybinds. I think this follows other compositors' behaviors the most. A downside of this approach is, obviously, Firefox still shows its menu bar when other actions like `Execute` are triggered with Alt-* keybinds. KWin has this problem, but it's less problematic than labwc since its default keybinds are mostly triggered with Win key rather than Alt key and users are used to that.

   This can be tested here: https://github.com/tokyo4j/labwc/tree/always-send-modifier-release

My personal preference is (1.) since I think it's the least impactful solution, but I'm not sure.

The motivation for opening this discussion is that I recently received a personal email from 60IQ (whose GitHub account was deleted for some reason) that explains the same issue as Blender's but now with xwayland games like gzdoom. Also I found #1537 which I guess is also the same problem as Blender's. They're confusing at a glance and I don't think forcing users to use `wantAbsorbedModifierReleaseEvents` window rule to fix it window-by-window is good for the out-of-the-box experience of labwc.

**Comment by Consolatis** [comment]
Personally I was never a big fan of sending a press but no release event. If the modifier_press -> keyboard.leave solution works in practice to reset the modifier state on re-enter (including with blender) that sounds like a better solution to me. The key-state tracking is likely the area of labwc I have the least amount of knowledge about so I might miss something.

CC @jlindgren90 @johanmalm

**Comment by jlindgren90** [comment]
Sending wl_keyboard.{leave,enter} sounds more like what happens under X11. If this works well in practice, I'm all for it.

Has anyone check what kwin does, for a comparison?

**Reply by tokyo4j** [reply]
Kwin does nothing special. It always sends release events for modifier keys when they're released, so firefox shows its menu bar when the user registers Alt-* keybind and uses it. Activating window-switcher doesn't suffer from it though, because it sends `wl_keyboard.leave` when window-switcher is activated. Other than Alt-Tab, KWin's default keybinds are mostly bound to Win key rather than Alt key, so I guess most users don't complain about it.

**Comment by tokyo4j** [comment]
Opened #2452.

**Comment by johanmalm** [comment]
Thanks for working on this and apologies for being slow to respond. I knew this would take a bit more brainpower than I could probably summon over the local festive holiday.

I agree that the use of `wantAbsorbedModifierReleaseEvents` is a poor solution and I'm very supportive of trying something that gets rid of it.

I think it can be hard to evaluate these key-state solutions until they hit users. When it comes to keybinds and which events are sent to clients, I have many times thought I had it implemented correctly and only later - when users used it - realised that there were issues.

Intuitively it feels like there could be unintended consequences with option 1 (sending the `wl_keyboard.{leave,enter}` events).

> The more I consider this problem, the more I think we should take the option (2) in #2448: just send a leave event on window-switcher activation. This only works for the default Alt-Tab keybind and will reintroduce the firefox's menu bar issue with some existing keybinds like: `<keybind key="A-Left"><action name="MoveToEdge" direction="left" /></keybind>`.
> But Kwin also has this problem and I think it's a much more minor issue than that with Alt-Tab keybind. As we advertise labwc as a lightweight and simple compositor, I prefer this way.

I agree.

We could consider phasing out the default `alt-` keybinds. They are just some 'random' ones that I chose right at the start of labwc development. I don't take this lightly because it can be upsetting to users when these things change, but there is a time/place to choose the less ugly and actually do things right. I personally remap them all to use the `super-` key anyway so there is some irony in this :smile:

Do you want me to do some testing with the `always-send-modifier-release` branch or is it still work-in-progress?

**Reply by Consolatis** [reply]
> We could consider phasing out the default `alt-` keybinds. They are just some 'random' ones that I chose right at the start of labwc development. I don't take this lightly because it can be upsetting to users when these things change, but there is a time/place to choose the less ugly and actually do things right.

I see it similar for all keybinds other than `A-Tab` and `A-F4`. This includes the mousebinds like replacing A-Drag-Left for moving windows with `W-Drag-Left` by default, same for the SnapToEdge ones. We could start by only adding the `W-` variants for now and then at some point slowly start to deprecate and finally remove the `A-` based defaults.

**Comment by tokyo4j** [comment]
@johanmalm @jlindgren90

I could finally fix my broken GNOME environment and see how Mutter handles this. The result was surprising; Mutter sends release events of all the keys even when it triggered a keybinding.

For example, when I press Alt-F10 to toggle maximization state of the focused window in Mutter, only the press event of F10 key is absorbed by Mutter and its corresponding release event is sent to the client. Here's the comparison of how Mutter, KWin and labwc handle this:

|keyboard signals|Mutter|KWin|labwc|
|-|-|-|-|
|Alt press|sent to client|sent to client|sent to client
|F10 press|absorbed|absorbed|absorbed|
|F10 release|sent to client|absorbed|absorbed|
|Alt release|sent to client|sent to client|absorbed|
||||
|result|works perfectly|shows menu bar in firefox|Alt key is stuck in blender|

And surprisingly, Firefox doesn't show its menu bar if the release event of F10 is sent. I'm not sure if this behavior is universal among all the application, but I feel this Mutter's behavior is very appealing to us, because it won't degrade our users' experience by showing Firefox's menu bar when Alt-* keybind is triggered.

You can try it with a very simple patch:

```diff
diff --git a/src/input/keyboard.c b/src/input/keyboard.c
index cef86bbf..86bc42b7 100644
--- a/src/input/keyboard.c
+++ b/src/input/keyboard.c
@@ -415,7 +415,7 @@ handle_key_release(struct server *server, uint32_t evdev_keycode,
 	 * If a press event was handled by a compositor binding, then do
 	 * not forward the corresponding release event to clients.
 	 */
-	return true;
+	return false;
 }
 
 static bool

```

Do you think this is the right approach for us? Or do you think this is just an odd behavior of Mutter and KWin's behavior (not sending the release of F10) is better?

> Do you want me to do some testing with the always-send-modifier-release branch or is it still work-in-progress?

~So let's discuss the approach I propose above first.~ Now that I realized the proposed solution will have issues with Amberol (see my reply below), I'll open a PR for the always-send-modifier-release later. Then you can test it there.

**Reply by jlindgren90** [reply]
I'm good with trying this - imitate mutter's behavior. @johanmalm do you recall if this was tried already and whether it caused any issues (if so)?

**Reply by tokyo4j** [reply]
Hmm.... I found [an issue in Mutter](https://gitlab.gnome.org/GNOME/mutter/-/issues/2638) caused by the Mutter's behavior that always sends key release events. Looks like Amberol listens for key __releases__ rather than key presses for playback/playforward. So my suggestion doesn't seem to be the best option and the option (2.) sounds more robust.

Sorry for going back and forth between ideas.

**Reply by jlindgren90** [reply]
Good find. The discussion there makes it sound like the mutter behavior is more of an implementation accident than a conscious choice.

> Sorry for going back and forth between ideas.

We have all been doing the same. No need to apologize :)

**Reply by johanmalm** [reply]
Well - I haven't documented it, but I'm fairly sure that passing `absorbed` non-modifier key release events (F10 release) to clients causes problems with some application. That was the original idea behind introducing `key-state.c`. 
I can't remember the details, but between gtk2/3/4 & qt5/6 across XWayland and Wayland - as well as Tk, Java, etc the only consistent approach was to not pass the absorbed key events. This is also what sway does. 
It is mysterious to me that mutter does not do this. Not sure what's going on there, but don't feel we should copy that one :smile:
